### DISCUSSION TITLE
In the TDOA algorithm, can the distance value of the two base stations be measured by a precise instrument to replace the distance measurement value of the TWR method in the program.

### DISCUSSION THREAD

**Discussion by lalalalang999** [bitcraze/discussions#361] [Q&A] [answered]
hi.
In the TDOA2 algorithm, I saw that the program uses the TWR method to measure the distance between the two base stations. In order to improve the accuracy of TDOA, can I use a high-precision instrument to directly measure the distance between the two base stations? instead of the dw1000 measurement.
In the TDOA2 algorithm, the distance value of the two base stations is calculated 
[here](https://github.com/bitcraze/lps-node-firmware/blob/2fe572e4275b1fbb11937c45ae74bb8ff7eb4ee0/src/uwb_tdoa_anchor2.c#L195)。
But my actual attempt is incorrect, for example, the value of ctx.distances[slot] base station 0 and base station 1 is 34870. Calculated by dw1000 clock:
34870 * speed of light * 1/(128*499.2e6) = 163.6018m
The actual distance between base station 0 and base station 1 is 8.8726m. Where am I wrong?

**Comment by krichardsson** [comment, marked as answer]
Hi!

> In order to improve the accuracy of TDOA, can I use a high-precision instrument to directly measure the distance between the two base stations?

The short answer is no, it will make it worse. 

Now to the longer explanation. To understand this problem, let's imagine that we have an ideal system where we have a shared clock in all devices (the two anchors and the crazyflie) and where the UWB radios can receive multiple packets at the same time. The easiest way to get the TDoA in the Crazyflie would be if both the anchors transmitted a packet at exactly the same time, then all we have to do in the Crazyflie would be to measure the time between when the packets are received. Easy!

The real world is slightly more complicated and the main obstacles in this case are:
* The clocks in the devices are not synchronized (both phase and rate). However the rate changes slowly and by looking at the time stamps of received packets, we can calculate the ratio between the rate of our local clock and an external clock.
* The UWB radios can not receive multiple packets at the same time (one or both will be lost) and we can not receive at the same time as we transmit.

As we do not know the phase of other clocks, it is not possible to make two devices transmit at a specific time based on their clocks so we need a different mechanism for this. The solution is to transmit a packet from one of the anchors (let's call the anchor A), this is one of the packets in the example above and the reception time will be recorded when received by the Crazyflie (trxA). This packet is also received by the other anchor (anchor B) which in turn triggers a transmission of a new packet, this is the other packet in the ideal world example and the receive time (trxB) is recorded in the Crazyflie. Obviously there is a delay between the reception and transmission of the packets in anchor B, luckily this can be measured using the local UWB clock (let's call it td). Assume the time it takes for a packet to go from anchor A to anchor B is tAtoB, then we can calculate the TDoA: `tdoa = (trxB - td - tAtoB) - trxA`
This is exactly what the TWR between anchors is providing, the time of flight between anchor A and B. The key here is that we are interested in the **time**, not the distance. We want to know exactly how much later we transmitted from anchor B compared to anchor A, the distance is not relevant.

> But my actual attempt is incorrect, for example, the value of ctx.distances[slot] base station 0 and base station 1 is 34870. Calculated by dw1000 clock:
34870 * speed of light * 1/(128*499.2e6) = 163.6018m
The actual distance between base station 0 and base station 1 is 8.8726m. Where am I wrong? 

The reason you get weird numbers is called the antenna delay. There is a bit of black magic in radio and it takes a while for the radio waves to propagate through the antenna (and other parts of the system), quite a long time actually, converted to meters (using the speed of light) it is around [154.6](https://github.com/bitcraze/crazyflie-firmware/blob/d89b96226ddd4540171770a036a375c8606b4f1f/src/deck/drivers/interface/locodeck.h#L47) meters. This number includes all "non air time" from transmitter to receiver. In your case you would get `163.6 - 154.6 = 8.7`, not too far away from the expected 8.9m.

I will add a bit more information on the antenna delay (this is becoming a long answer :-)). In all our code we assume the antenna delay is constant, which unfortunately is not true (and probably one of the reasons to the 0.2m error in your measurement). The delay depends on all sorts of things but the two main problems (as far as I understand) is antenna orientation and the environment around the device. I have tried to measure the antenna delay in various angles and environments and the result varies quite a lot, up to 0.5m! I don't think it is possible to create a simple function to compensate for it. Wenda Zhao et al has worked on this problem, you can find a paper here: https://arxiv.org/abs/2103.01885v1 and a data set here: https://github.com/utiasDSL/util-uwb-dataset.

**Reply by lalalalang999** [reply]
hi！
Sorry for taking so long to reply.Yes, the antenna delay has a very big impact on UWB ranging. As set in the program the antenna delay is #define ANTENNA_OFFSET 154.6.
 However, I see in the program that this antenna delay is not used under the TDOA2 algorithm. It is used in both TDOA3 and TWR algorithms. What I am wondering is that the TDOA2 algorithm does not need antenna delay calibration? Or is it possible to do the antenna delay, but it is just not added to the TDOA2 algorithm.
Also. I read the DW1000 datasheet and know that there exists an OTP area where the antenna delay data can be stored. My idea is to measure the antenna delay. Initially configure it into the register of DW1000, whether this can improve the accuracy of TDOA2 algorithm.

Yes. I am more interested in the TDOA2 algorithm than the TDOA3 algorithm. This is because the TDOA2 algorithm has better accuracy than the TDOA3 algorithm. I would also like to further improve the accuracy of the TDOA2 algorithm by adding antenna delays.

**Reply by krichardsson** [reply]
>  see in the program that this antenna delay is not used under the TDOA2 algorithm. It is used in both TDOA3 and TWR algorithms. What I am wondering is that the TDOA2 algorithm does not need antenna delay calibration? Or is it possible to do the antenna delay, but it is just not added to the TDOA2 algorithm.

The antenna delay cancels out (sort of) in the TDoA algorithms and is not needed in TDoA2 or TDoA3. You say that it is used in TDoA3, where did you find that?

The antenna delay is canceled out in TDoA algorithms based on the assumption that it is the same delay when receiving from both anchors. The problem is that this is not true (as you have noted) and to improve the situation you would have to figure out the delay based on the CF position/orientation, anchor orientations, obstacles, the environment and so on. This is a very complex task and I have not seen any analytical solutions to this problem. Let us know if you succeed!
